core/corehttp/gateway_proxy.go

```diff
@@ -53,6 +53,10 @@ func ProxyOption() ServeOption {
 	return func(n *core.IpfsNode, l net.Listener, mux *http.ServeMux) (*http.ServeMux, error) {
 		childMux := http.NewServeMux()
 		mux.HandleFunc("/", func(w http.ResponseWriter, r *http.Request) {
+			// TODO remove below dead code?
+			// Rationale Go does not support requests with CONNECT method :-(
+			// https://golang.org/src/net/http/request.go#L111
+			// Below code block never gets executed
 			if r.Method == http.MethodConnect {
```

Comment:
Unless I got it totally wrong, [request handling in `http` does not really support `CONNECT` method](https://golang.org/src/net/http/request.go#L111). 

Based on my debug the inside of this `if` in `ProxyOption` is never used, even when `CONNECT` is forced via `curl --proxytunnel -x`.

It is not a problem because regular HTTP proxy (`-x` without `--proxytunnel` in `curl`) works as expected, and we know that `CONNECT`-based proxy tunneling  won't be used for local subdomains in web browsers.
One liner for proxy testing in browser:
```console
$ chromium --incognito --proxy-server="http://127.0.0.1:8030" "http://bafybeiemxf5abjwjbikoz4mc3a3dla6ual3jsgpdr4cjr3oz3evfyavhwq.ipfs.localhost:8030"
```

To keep the surface of this PR minimal I propose we remove `ProxyOption` and limit its scope to regular HTTP proxy mode (which already works and got sharness tests). 

@Stebalien Thoughts?

Reply:
Go's http _client_ does not support _sending_ requests with the CONNECT method. If you want to use a proxy, you have to specify a `Proxy` function on the transport (https://golang.org/pkg/net/http/#Transport).

Reply:
I wonder if something else is hijacking the request? Or handling it first?

Reply:
Indeed. All `CONNECT` requests retuned HTTP 404. They were handled by `http.Server`, but no `ServeMux`-based handler was capable of acting on it (they have similar limitation as http client). 

Not sure if this is the most elegant way, but I've created a thin wrapper that checks HTTP method before passing request to `ServeMux` handlers, and if it is `CONNECT` it just returns HTTP 200 to signal to the client that regular proxy GET can be issued:

https://github.com/ipfs/go-ipfs/commit/bcb3a9552b89cd113287697e5caab4ddd5c39b06#diff-6eb71e82338561cf3806f670b1156365R46

For what its worth, I enabled tests for `CONNECT` mode and it all works as expected.

Reply:
Interesting. That's much simpler than the old approach. LGTM.